[ray_client] actors v0

barakmich opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

Depends on #12141 
Closes #11976


## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ericl commented:
Needs rebase / pull of master.

barakmich commented:
Of course it does. I mean, I did say "Depends on #12141" -- which has now been merged

ericl commented on python/ray/experimental/client/client_app.py:
```diff
@@ -3,6 +3,27 @@
+actor = HelloActor.remote()
+s, count = ray.get(actor.say_hello.remote("you"))
+print(s, count)
+assert s == "hello you"
```
```suggestion
assert s == "Hello you"
```

AmeerHajAli replied:
Is this passing?

ericl commented on python/ray/experimental/client/client_app.py:
```diff
@@ -3,6 +3,27 @@
+assert count == 1
+s, count = ray.get(actor.say_hello.remote("world"))
+print(s, count)
+assert s == "hello world"
```
```suggestion
assert s == "Hello world"
```

ericl commented on python/ray/experimental/client/client_app.py:
```diff
@@ -3,6 +3,27 @@
+s, count = ray.get(actor.say_hello.remote("world"))
+print(s, count)
+assert s == "hello world"
+assert count == 2
```
One thing I noticed is that passing actor handles crashes trying to serialize the handle:

```
@ray.remote
def f(handle):
    return handle.say_hello.remote("Bob")

ray.get(f.remote(actor))
```

ericl commented on python/ray/experimental/client/common.py:
```diff
@@ -27,12 +38,64 @@ def __call__(self, *args, **kwargs):
+        self.actor_class = actor_class
+
+    def __getattr__(self, key):
+        return ClientRemoteMethod(self, key)
```
It seems __getattr__ is intercepting too much here, e.g., if you try to pickle this handle it ends up calling __getattr__ which returns a remote method.

ericl reviewed (commented):
Just tried it out, works well. Found a corner case with the getattr handling ^^

AmeerHajAli commented on python/ray/experimental/client/client_app.py:
```diff
@@ -3,6 +3,27 @@
+    def __init__(self):
+        self.count = 0
+
+    def say_hello(self, whom):
```
type annotation please?

AmeerHajAli commented on python/ray/experimental/client/server/server.py:
```diff
@@ -68,24 +71,62 @@ def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
             if not isinstance(func, ClientRemoteFunc):
-                raise Exception("Attempting to schedule something that "
+                raise Exception("Attempting to schedule function that "
                                 "isn't a ClientRemoteFunc")
```
```suggestion
                                "isn't a ClientRemoteFunc.")
```

AmeerHajAli commented on python/ray/experimental/client/server/server.py:
```diff
@@ -68,24 +71,62 @@ def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
+                actor_class = ray.get(actor_class_ref)
+                if not inspect.isclass(actor_class):
+                    raise Exception("Attempting to schedule actor that "
+                                    "isn't a ClientActorClass")
```
```suggestion
                                    "isn't a ClientActorClass.")
```

AmeerHajAli commented on python/ray/experimental/client/server/server.py:
```diff
@@ -68,24 +71,62 @@ def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
+            self.object_refs[output.binary()] = output
+        return ray_client_pb2.ClientTaskTicket(return_id=output.binary())
+
+    def _schedule_actor(self, task, context=None):
```
type annotation?

AmeerHajAli commented on python/ray/experimental/client/worker.py:
```diff
@@ -96,21 +102,62 @@ def wait(self,
+            raise TypeError("The @ray.remote decorator must be applied to "
+                            "either a function or to a class.")
+
+    def call_remote(self, instance, kind, *args, **kwargs):
```
type annotation?

AmeerHajAli reviewed (commented):
LGTM.
I am able to reproduce the bug Eric mentioned on passing an actor handler to a function.
I also tested a bunch of other combinations of arguments (e.g., passing functions/reference as arguments) but these seem to work. What also does not work is passing an actor handler to an actor (to any function in the actor).
Can we have another test in addition to the basic one that tests a bunch of combinations of passing references/functions/actors?

barakmich commented:
Addressed and rebased -- passing an ActorHandle around is kind of bleh. I mean, I'm forming a plan think I can do it (and, similarly, functions) in a reasonable way by tomorrow but that's some 'fun' behavior.

ericl commented:
@barakmich is there something here beyond fixing the getattr problem? I believe the way normal Ray actor handles work is by building an actor method table. We can probably do something similar to override just those methods instead of getattr, which is preventing serialization of those handles.

ericl commented:
@AmeerHajAli what do you think about merging this as part1 of actors--- fixing actor handles can be part2

ericl commented:
I think we can merge this as part 1 of actors... lint is failing though:
```

++++command -v clang-format
++++/home/travis/build/ray-project/tmp.sv31RqRIQV/ci/travis/check-git-clang-format-output.sh
Running clang-format against branch 234df9091e57f2a4d7574e2ae53635bb5c862b9e, with hash 234df9091e57f2a4d7574e2ae53635bb5c862b9e
clang-format failed:
diff --git a/src/ray/protobuf/ray_client.proto b/src/ray/protobuf/ray_client.proto
index 7f70460c5..fd8fe5345 100644
--- a/src/ray/protobuf/ray_client.proto
+++ b/src/ray/protobuf/ray_client.proto
@@ -29,7 +29,6 @@ message Arg {
   Type type = 4;
 }```

AmeerHajAli commented:
We can split it to two PRs, can we open an issue for this though so we can make sure to track it properly?

barakmich commented:
SGTM,  found the lint that only `--all` caught

ericl commented:
Looks good, let's merge and iterate on the fixes
